### PR TITLE
[BARX-1656] Add registryMigrationMode for gradual migration to registry.datadoghq.com

### PR DESCRIPTION
**Pull request by AliDatadog**
#### What this PR does / why we need it:

Adds `registryMigrationMode` to the datadog chart to control gradual migration of the default image registry from site-specific mirrors (GCR, ACR) to `registry.datadoghq.com`.

A new `registryMigrationMode` setting controls the behavior:
- `"auto"` (default): enables `registry.datadoghq.com` for sites where migration is rolled out. Currently enabled: **AP1** (`ap1.datadoghq.com`) when `datadog.apm.enabled` is `false` (the default). More sites and features will be enabled in future releases.
- `"all"`: enables `registry.datadoghq.com` for all eligible sites (AP1, AP2, EU1, US1, US5).
- `""`: disables migration, keeping site-specific registries.

**Exclusions (regardless of mode):**
- **US3** (`us3.datadoghq.com`): always uses `datadoghq.azurecr.io`
- **US1-FED** (`ddog-gov.com`): always uses `public.ecr.aws/datadog`
- **GKE Autopilot** clusters: always use `gcr.io` variants
- **GKE GDC** clusters: always use `gcr.io` variants
- Explicit `registry` value always takes precedence.

A **NOTICE** is displayed during `helm install/upgrade` when registry migration is active for the current site, with instructions to revert if needed.

This is the first step of the [registry.datadoghq.com k8s rollout schedule](https://datadoghq.atlassian.net/wiki/spaces/BARX/pages/6314722847). Future PRs will progressively enable more sites in `"auto"` mode.

#### Which issue this PR fixes

- Part of https://datadoghq.atlassian.net/browse/BARX-1655

#### Special notes for your reviewer:

Similar to the operator chart's `registryMigrationMode` in #2421, but applied directly in the `registry` helper template instead of through env vars.

The `registry` helper in `_helpers.tpl` now checks `registryMigrationMode` before falling back to site-specific registries. The logic is:
1. If `registry` is set → use it (unchanged)
2. If `ddog-gov.com` → ECR (unchanged)
3. If `us3.datadoghq.com` → ACR (unchanged)
4. If migration is active for this site AND not GKE Autopilot/GDC → `registry.datadoghq.com`
5. Otherwise → site-specific registry (unchanged)

**APM gating for AP1:** In `"auto"` mode, AP1 only migrates when `datadog.apm.enabled` is `false` (the default). If a customer has explicitly enabled APM (`datadog.apm.enabled: true`), AP1 remains on `asia.gcr.io/datadoghq`.

Tests cover all 7 Datadog sites (US1, US3, US5, EU1, US1-FED, AP1, AP2) × 3 modes (auto/all/disabled), plus APM condition, GKE Autopilot, GKE GDC, and explicit registry override edge cases (28 test cases total).

#### Checklist
- [x] All commits are signed (see: [signing commits][1])
- [x] Chart Version semver bump label has been added (use `<chartName>/minor-version`, `<chartName>/patch-version`, or `<chartName>/no-version-bump`)
- [x] For `datadog` or `datadog-operator` chart or value changes, update the test baselines (run: `make update-test-baselines`)

GitHub CI takes care of the below, but are still required:
- [x] Documentation has been updated with helm-docs (run: `.github/helm-docs.sh`)
- [x] `CHANGELOG.md` has been updated
- [x] Variables are documented in the `README.md`

[1]: https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

### CODE REVIEW

**Comment by AliDatadog** [comment]
@codex review

**Comment by AliDatadog** [comment]
@codex review

File: charts/datadog/templates/_helpers.tpl [hunk 1]
```diff
@@ -458,18 +458,31 @@ Return the proper registry based on datadog.site (requires .Values to be passed
...
+{{- $migrationMode := default "" .registryMigrationMode -}}
+{{- $migratedSite := false -}}
+{{- if eq $migrationMode "all" -}}
+{{- $migratedSite = true -}}
+{{- else if eq $migrationMode "auto" -}}
+{{- if eq $site "ap1.datadoghq.com" -}}
+{{- $migratedSite = true -}}
+{{- end -}}
+{{- end -}}
+{{- if and $migratedSite (not .providers.gke.autopilot) -}}
```
Comment (fanny-jiang) [inline charts/datadog/templates/_helpers.tpl]: We'll need to exclude GKE GDC environments as well


```suggestion
{{- if and $migratedSite (not (or .providers.gke.autopilot .providers.gke.gdc)) -}}
```

File: charts/datadog/templates/NOTES.txt [hunk 2]
```diff
@@ -945,3 +945,31 @@ For better security, consider using an existing secret by setting:
...
+{{- $migrationMode := default "" .Values.registryMigrationMode -}}
+{{- $migratedSite := false -}}
+{{- if eq $migrationMode "all" -}}
+{{- $migratedSite = true -}}
+{{- else if eq $migrationMode "auto" -}}
+{{- if eq $site "ap1.datadoghq.com" -}}
+{{- $migratedSite = true -}}
+{{- end -}}
+{{- end -}}
+{{- if and $migratedSite (not .Values.registry) (ne $site "ddog-gov.com") (not .Values.providers.gke.autopilot) }}
```
Comment (fanny-jiang) [inline charts/datadog/templates/NOTES.txt]:
```suggestion
{{- if and $migratedSite (not .Values.registry) (ne $site "ddog-gov.com") (not (or .Values.providers.gke.autopilot .Values.providers.gke.gdc)) }}
```

File: charts/datadog/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Datadog changelog
 
+## 3.182.0
+
+* Add `registryMigrationMode` to control gradual migration of default image registry to `registry.datadoghq.com`, replacing site-specific regional mirrors (GCR, ACR). Defaults to `"auto"`, which currently enables `registry.datadoghq.com` for the AP1 site (`ap1.datadoghq.com`). More sites will be enabled in future releases. Set to `""` to disable. GKE Autopilot, US1-FED, and US3 clusters are excluded.
```
Comment (fanny-jiang) [inline charts/datadog/CHANGELOG.md]:
```suggestion
* Add `registryMigrationMode` to control gradual migration of default image registry to `registry.datadoghq.com`, replacing site-specific regional mirrors (GCR, ACR). Defaults to `"auto"`, which currently enables `registry.datadoghq.com` for the AP1 site (`ap1.datadoghq.com`). More sites will be enabled in future releases. Set to `""` to disable. GKE Autopilot, GKE GDC, US1-FED, and US3 clusters are excluded.
```

File: charts/datadog/README.md [hunk 4]
```diff
@@ -1109,7 +1109,8 @@ helm install <RELEASE_NAME> \
...
-| registry | string | `nil` | Registry to use for all Agent images (default to [gcr.io | eu.gcr.io | asia.gcr.io | datadoghq.azurecr.io | public.ecr.aws/datadog] depending on datadog.site value) |
+| registry | string | `nil` | Registry to use for all Agent images (default depends on datadog.site and registryMigrationMode values) |
+| registryMigrationMode | string | `"auto"` | Controls gradual migration of default image registry to registry.datadoghq.com, replacing site-specific regional mirrors (GCR, ACR). This setting has no effect when `registry` is explicitly set. GKE Autopilot clusters are excluded and always use gcr.io/datadoghq. US1-FED (ddog-gov.com) is excluded and always uses public.ecr.aws/datadog. US3 (us3.datadoghq.com) is excluded and always uses datadoghq.azurecr.io. |
```
Comment (fanny-jiang) [inline charts/datadog/README.md]:
```suggestion
| registryMigrationMode | string | `"auto"` | Controls gradual migration of default image registry to registry.datadoghq.com, replacing site-specific regional mirrors (GCR, ACR). This setting has no effect when `registry` is explicitly set. GKE Autopilot and GKE GDC clusters are excluded and always use gcr.io/datadoghq. US1-FED (ddog-gov.com) is excluded and always uses public.ecr.aws/datadog. US3 (us3.datadoghq.com) is excluded and always uses datadoghq.azurecr.io. |
```

File: charts/datadog/values.yaml [hunk 5]
```diff
@@ -35,6 +35,19 @@ commonLabels: {}
 
 registry:  # gcr.io/datadoghq
 
+# registryMigrationMode -- Controls gradual migration of default image registry to
+# registry.datadoghq.com, replacing site-specific regional mirrors (GCR, ACR).
+# This setting has no effect when `registry` is explicitly set.
+# GKE Autopilot clusters are excluded and always use gcr.io/datadoghq.
```
Comment (fanny-jiang) [inline charts/datadog/values.yaml]:
```suggestion
# GKE Autopilot and GKE GDC clusters are excluded and always use gcr.io/datadoghq.
```

File: charts/datadog/templates/_helpers.tpl (hunk 1 above)
Comment (AliDatadog) [reply charts/datadog/templates/_helpers.tpl]: Done ! Thank you

**Comment by AliDatadog** [comment]
@codex review

File: charts/datadog/templates/_helpers.tpl [hunk 6]
```diff
@@ -458,18 +458,34 @@ Return the proper registry based on datadog.site (requires .Values to be passed
 {{- $site := default "datadoghq.com" .datadog.site -}}
 {{- if .registry -}}
 {{- .registry -}}
-{{- else if eq $site "datadoghq.eu" -}}
-eu.gcr.io/datadoghq
 {{- else if eq $site "ddog-gov.com" -}}
 public.ecr.aws/datadog
+{{- else if and (eq $site "us3.datadoghq.com") (not .providers.gke.autopilot) (not .providers.gke.gdc) -}}
```
[the remark this answers is not in the document]
Comment (fanny-jiang) [reply charts/datadog/templates/_helpers.tpl:467]: @AliDatadog In addition to addressing the above, can we also please the new `registryMigrationMode` as a property in the [values.schema.json](https://github.com/DataDog/helm-charts/blob/main/charts/datadog/values.schema.json)? 

example:
```
{
  "properties": {
    "registryMigrationMode": {
      "type": "string",
      "enum": ["", "auto", "all"],
      "description": "Controls gradual migration of the default image registry to registry.datadoghq.com."
    }
  }
}
```

[the remark this answers is not in the document]
Comment (AliDatadog) [reply charts/datadog/templates/_helpers.tpl:467]: Done ! Thanks

**Comment by AliDatadog** [comment]
@codex review

File: charts/datadog/README.md [hunk 7]
```diff
@@ -1109,7 +1109,8 @@ helm install <RELEASE_NAME> \
...
-| registry | string | `nil` | Registry to use for all Agent images (default to [gcr.io | eu.gcr.io | asia.gcr.io | datadoghq.azurecr.io | public.ecr.aws/datadog] depending on datadog.site value) |
+| registry | string | `nil` | Registry to use for all Agent images (default depends on datadog.site and registryMigrationMode values) |
+| registryMigrationMode | string | `"auto"` | Controls gradual migration of default image registry to registry.datadoghq.com, replacing site-specific regional mirrors (GCR, ACR). This setting has no effect when `registry` is explicitly set. GKE Autopilot and GKE GDC clusters are excluded and always use gcr.io/datadoghq. US1-FED (ddog-gov.com) is excluded and always uses public.ecr.aws/datadog. US3 (us3.datadoghq.com) is excluded and always uses datadoghq.azurecr.io. |
```
[the remark this answers is not in the document]
Comment (AliDatadog) [reply charts/datadog/README.md]: Safe to ignore
